### PR TITLE
[Reviewer: Andy] Enhance Chef to do one-shot scale-up

### PR DESCRIPTION
**Pull request by mirw**
Andy,

Please can you review my enhancement to chef to support Astaire and use it to do one-shot scale-up/down.  The old two-phase scale-up/down is still supported.

I've live tested (as you've seen).

Matt

### CODE REVIEW

File: plugins/knife/cluster-boxes.rb [hunk 1]
```diff
@@ -68,5 +68,54 @@ def add_cluster_role(role)
...
+
+    # Run the specified `astaire` command on all nodes in the local environment
+    # that match the given `query_string`.
+    #
+    # @param cloud [Symbol] The cloud hosting the devices.
+    # @param query_string [String] A Chef-format query string to match on.
+    # @param command [String] The command to send to `astaire`.
+    def run_astaire(cloud, command)
+      %w{sprout ralf}.each do |role|
+        run_command(cloud, query_string(true, role: role), "sudo service astaire #{command}")
```
Comment (mirw) [inline plugins/knife/cluster-boxes.rb]: If there aren't any ralf nodes, this fails with a `FATAL: No nodes returned from search!` error.  Do you know how to check whether there are any matching results before doing this (or not return a fatal error if none are found)?

File: plugins/knife/knife-deployment-resize.rb [hunk 2]
```diff
@@ -359,45 +363,179 @@ def calculate_box_counts(config)
...
         end
 
-        if config[:force] or still_quiescing.empty?
-          # Safe to delete quiesced boxes.
-          Chef::Log.info "Deleting quiesced boxes..."
-          delete_quiesced_boxes env
-        else
-          Chef::Log.error "#{still_quiescing} are still quiescing, can't finish (use --force to force it at the risk of data loss or call failures)'"
+        if !config[:force]
+          # Make Astaire wait for synchroniztion to complete.
```
Comment (bossmc) [inline plugins/knife/knife-deployment-resize.rb]: sp: "synchroniztion"

File: plugins/knife/cluster-boxes.rb [hunk 3]
```diff
@@ -68,5 +68,54 @@ def add_cluster_role(role)
...
+    # @param query_string [String] A Chef-format query string to match on.
+    def trigger_chef_client(cloud, query_string)
+      run_command(cloud, query_string, "sudo nice -n 19 chef-client")
+    end
+
+    # Run the specified `astaire` command on all nodes in the local environment
+    # that match the given `query_string`.
+    #
+    # @param cloud [Symbol] The cloud hosting the devices.
+    # @param query_string [String] A Chef-format query string to match on.
```
Comment (bossmc) [inline plugins/knife/cluster-boxes.rb]: This param doesn't exist.

File: plugins/knife/cluster-boxes.rb [hunk 4]
```diff
@@ -68,5 +68,54 @@ def add_cluster_role(role)
...
+      if cloud == :openstack
+        # Guard against boxes which do not have a public hostname
+        knife_ssh.config[:attribute] = 'ipaddress'
+      end
+      knife_ssh.config[:identity_file] = "#{attributes["keypair_dir"]}/#{attributes["keypair"]}.pem"
+      knife_ssh.config[:verbosity] = config[:verbosity]
+      Chef::Config[:verbosity] = config[:verbosity]
+      knife_ssh.config[:on_error] = :raise
+      # Run chef-client at maximum niceness to minimize the hit on potentially
+      # heavily loaded nodes.
```
Comment (bossmc) [inline plugins/knife/cluster-boxes.rb]: Outdated comment.

File: plugins/knife/cluster-boxes.rb (hunk 1 above)
Comment (bossmc) [inline plugins/knife/cluster-boxes.rb]: Apparently not (see https://github.com/chef/chef/blob/master/lib/chef/knife/ssh.rb#-145-148)... you can check for the presence of the nodes yourself if you care (https://github.com/Metaswitch/chef/blob/master/plugins/knife/clearwater-dns-records.rb#-110-122).

**Comment by bossmc** [comment]
Looks good.

**Pull request merged**